troubleshooting/checkpointing/src/solution/java/com/ververica/flink/training/solutions/CheckpointingJobSolution3.java

```diff
@@ -0,0 +1,297 @@
+			 * f2: the value of the previous measurement
+			 */
+			return new Tuple3<>(0L, 0.0, 0.0);
+		}
+
+		@Override
+		public Tuple3<Long, Double, Double> add(
+				final Measurement record,
+				final Tuple3<Long, Double, Double> aggregate) {
+
+			if (aggregate.f0 > 0) {
+				aggregate.f1 += record.getValue() - aggregate.f2;
+			}
+			aggregate.f0++;
+			aggregate.f2 = record.getValue();
+			return aggregate;
+		}
+
+		@Override
+		public Tuple2<Long, Double> getResult(final Tuple3<Long, Double, Double> windowedMeasurements) {
+			return new Tuple2<>(windowedMeasurements.f0, windowedMeasurements.f1);
+		}
+
+		@Override
+		public Tuple3<Long, Double, Double> merge(
+				final Tuple3<Long, Double, Double> agg1,
+				final Tuple3<Long, Double, Double> agg2) {
+			// this way of aggregation does not support merging of two aggregator
+			throw new UnsupportedOperationException();
+		}
+	}
+
+	/**
+	 * Deserializes the JSON Kafka message.
+	 */
+	public static class MeasurementDeserializer extends
+			RichFlatMapFunction<FakeKafkaRecord, Measurement> {
+		private static final long serialVersionUID = 3L;
+
+		private Counter numInvalidRecords;
+		private transient ObjectMapper instance;
+
+		@Override
+		public void open(final Configuration parameters) throws Exception {
+			super.open(parameters);
+			numInvalidRecords = getRuntimeContext().getMetricGroup().counter("numInvalidRecords");
+			instance = createObjectMapper();
+		}
+
+		@Override
+		public void flatMap(final FakeKafkaRecord kafkaRecord, final Collector<Measurement> out) {
+			final Measurement node;
+			try {
+				node = deserialize(kafkaRecord.getValue());
+			} catch (IOException e) {
+				numInvalidRecords.inc();
+				return;
+			}
+			out.collect(node);
+		}
+
+		private Measurement deserialize(final byte[] bytes) throws IOException {
+			return instance.readValue(bytes, Measurement.class);
+		}
+	}
+
+	private static class MeasurementByTimeComparator implements Comparator<Tuple2<Measurement, Long>> {
+		@Override
+		public int compare(Tuple2<Measurement, Long> o1, Tuple2<Measurement, Long> o2) {
+			return Long.compare(o1.f1, o2.f1);
+		}
+	}
+
+	/**
+	 * Calculates data for retrieving the average temperature difference between two sensor readings
+	 * (in event-time order!).
+	 */
+	public static class MeasurementWindowProcessFunction
+			extends
+			ProcessWindowFunction<Tuple2<Long, Double>, WindowedMeasurements, Integer, TimeWindow> {
+		private static final long serialVersionUID = 1L;
+
+		private static final int EVENT_TIME_LAG_WINDOW_SIZE = 10_000;
+
+		private transient DescriptiveStatisticsHistogram eventTimeLag;
+
+		@Override
+		public void process(
+				final Integer sensorId,
+				final Context context,
+				final Iterable<Tuple2<Long, Double>> input,
+				final Collector<WindowedMeasurements> out) {
+
+			final TimeWindow window = context.window();
+			Tuple2<Long, Double> result = input.iterator().next();
```

Comment:
Do we wait to either
1. check that this is the only element, or
2. just assume that it is, based on the code above? (in that case, please add a comment)

Reply:
Isn't it always the same for all Flink jobs: the iterator contains only one element?  Or are you talking about something specific to this job?

Reply:
Well, ok, it's always like this when you have an aggregation function in front...let's keep it as is...